### PR TITLE
Arbitrum Curve yield optimizers

### PR DESCRIPTION
**Pull request by johnnyonline**

### CODE REVIEW

File: src/arbitrum/utils/FortressArbiSwap.sol [hunk 1]
```diff
@@ -326,7 +326,7 @@ contract FortressArbiSwap is ReentrancyGuard, IFortressSwap {
         if (_toETH) {
             _unwrapETH(_amount);
         }
-
+
```
Comment (xh3b4sd) [inline src/arbitrum/utils/FortressArbiSwap.sol:329]: I guess we can do without these.

File: src/mainnet/concentrators/balancer/AuraBalConcentrator.sol [hunk 2]
```diff
@@ -66,60 +67,18 @@ contract AuraBalConcentrator is BalancerOperations, AMMConcentratorBase {
...
-        _updateRewards(_owner);
-
-        uint256 _assets = previewRedeem(_shares);
-        _withdraw(msg.sender, _receiver, _owner, _assets, _shares);
-
-        _withdrawStrategy(_assets, _receiver, false);
-
+    function _swapToUnderlying(address _underlyingAsset, uint256 _assets, uint256 _minAmount) internal override returns (uint256 _underlyingAmount) {
         _underlyingAmount = _removeLiquidity(address(asset), _underlyingAsset, _assets);
         if (!(_underlyingAmount >= _minAmount)) revert InsufficientAmountOut();
```
Comment (xh3b4sd) [inline src/mainnet/concentrators/balancer/AuraBalConcentrator.sol:79]: Is this method still supposed to swap to underlying? It looks like it only removes liquidity anymore without the swapping.

File: src/shared/compounders/AMMCompounderBase.sol [hunk 3]
```diff
@@ -60,7 +58,9 @@ abstract contract AMMCompounderBase is ReentrancyGuard, ERC4626 {
     address[] public rewardAssets;
     /// @notice The underlying assets.
     address[] public underlyingAssets;
-
+    /// @notice The internal accounting of AUM.
+    uint256 internal totalAUM;
```
Comment (xh3b4sd) [inline src/shared/compounders/AMMCompounderBase.sol:62]: Thinking about monitoring, can this kind of information be used by e.g. Dune so we can have public Fortress dashboards to show the evolution of AUM over time? Not sure right now if `totalAUM` is readable from the outside and I am also not sure what Dune for instance needs in order to query/display on chain information. Maybe just good for us to keep an eye on how certain information are meant to be exposed for on chain metrics.

File: src/shared/concentrators/AMMConcentratorBase.sol [hunk 4]
```diff
@@ -266,7 +265,29 @@ abstract contract AMMConcentratorBase is ReentrancyGuard, ERC4626 {
     /// @param _receiver - The receiver of minted shares.
     /// @param _minAmount - The minimum amount of assets (LP tokens) to receive.
     /// @return _shares - The amount of shares minted.
-    function depositSingleUnderlying(uint256 _underlyingAmount, address _underlyingAsset, address _receiver, uint256 _minAmount) external payable virtual nonReentrant returns (uint256 _shares) {}
+    // slither-disable-next-line reentrancy-no-eth
+    function depositSingleUnderlying(uint256 _underlyingAmount, address _underlyingAsset, address _receiver, uint256 _minAmount) external payable nonReentrant returns (uint256 _shares) {
+        if (!_isUnderlyingAsset(_underlyingAsset)) revert NotUnderlyingAsset();
+        if (!(_underlyingAmount > 0)) revert ZeroAmount();
+        
+        _updateRewards(msg.sender);
```
Comment (xh3b4sd) [inline src/shared/concentrators/AMMConcentratorBase.sol:273]: Is there a reason to "update rewards" for the user this early? Is there any implication to doing it here vs at the end? Any attack vectors here that we have to pay attention to?

File: test/arbitrum/compounders/curve/CurveCompounderBaseArbitrumTest.sol [hunk 5]
```diff
@@ -324,6 +324,7 @@ contract CurveCompounderBaseArbitrumTest is Test, AddressesArbi {
         vm.prank(harvester);
         uint256 _newUnderlying = curveCompounder.harvest(address(harvester), _asset, 0);
 
+        // From Curve dev discord "Arbitrum is probably a bit funkier -- I haven't dove into it, but I think it requires making a cross-chain call to trigger rewards that may be tough to reproduce"
         // assertTrue(IConvexBasicRewards(curveCompounder.crvRewards()).earned(address(curveCompounder)) == 0, "_testHarvest: E3");
```
Comment (xh3b4sd) [inline test/arbitrum/compounders/curve/CurveCompounderBaseArbitrumTest.sol:328]: Could you elaborate on this? What are the implications here? Why would we want this to work and what are the consequences now not to have it?

Comment (johnnyonline) [reply test/arbitrum/compounders/curve/CurveCompounderBaseArbitrumTest.sol:328]: As part of the `harvest` function tests, we push the forked blockchain about a month forward (denominated in blocks) to simulate the effects of having pending rewards for a vault.

In the current Curve/Convex Arbitrum setup, it seems that in order to update the _read_ amount of pending rewards, some cross-chain call needs to be made. The actual rewards (CRV tokens) are distributed as expected.

This doesnt effect the contracts, as the `isPendingRewards` is an external method.

File: src/mainnet/concentrators/balancer/AuraBalConcentrator.sol (hunk 2 above)
Comment (johnnyonline) [reply src/mainnet/concentrators/balancer/AuraBalConcentrator.sol:79]: I guess that `swap` is more of a terminology thing - the idea is to convert (swap/exchange/etc) the vault's primary asset into underlying assets, and vice versa.

Comment (johnnyonline) [reply src/mainnet/concentrators/balancer/AuraBalConcentrator.sol:79]: I guess that `swap` is more of a terminology thing - the idea is to convert (swap/exchange/etc) the vault's primary asset into underlying assets, and vice versa.

File: src/shared/compounders/AMMCompounderBase.sol (hunk 3 above)
Comment (johnnyonline) [reply src/shared/compounders/AMMCompounderBase.sol:62]: Certainly. `totalAUM` is readable using the `totalAssets` method.

File: src/shared/concentrators/AMMConcentratorBase.sol (hunk 4 above)
Comment (johnnyonline) [reply src/shared/concentrators/AMMConcentratorBase.sol:273]: Timing doesnt matter in this instance. This function violates the check-effect-interact pattern (which prevents against reentrancy attacks), but is using OZ's `nonReentrant` modifier.

We could use check-effect-interact pattern in this function, but we would need to assume that all the slippage the user is agreeing to have will actually happen (basically using `_minAmount` as `_assets`).

**Review by xh3b4sd — APPROVED** [review]
Go ahead as you see fit.

**Pull request merged**